documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,1189 @@
+    static virtual TVector Remainder(TVector left, TScalar right) => left % right;
+}
+
+public interface INumberVector<TVector, TScalar> :
+    IModulusVector<TVector, TScalar>
+    where TVector : INumberVector<TVector, TScalar>
+    where TScalar : INumber<TScalar>
+{
+    static abstract TVector CopySign(TVector value, TVector sign);
+    static abstract TVector CopySign(TVector value, TScalar sign);
+
+    static abstract TVector Sign(TVector value);
+}
+
+public interface IBinaryNumberVector<TVector, TScalar> :
+    INumberVector<TVector, TScalar>,
+    IBitwiseOperators<TVector, TVector, TVector>
+    where TVector : IBinaryNumberVector<TVector, TScalar>
+    where TScalar : IBinaryNumber<TScalar>
+{
+    static virtual TVector AllBitsSet => TVector.Create(TScalar.AllBitsSet);
+    
+    /// <summary>Computes the log2 of a value.</summary>
+    /// <param name="value">The value whose log2 is to be computed.</param>
+    /// <returns>The log2 of <paramref name="value" />.</returns>
+    static abstract TVector Log2(TVector value);
+}
+
+public interface IBinaryIntegerVector<TVector, TScalar> :
+    IBinaryNumberVector<TVector, TScalar>
+    where TVector : IBinaryIntegerVector<TVector, TScalar>
+    where TScalar : IBinaryInteger<TScalar>
+{
+    /// <summary>Computes the quotient and remainder of two values.</summary>
+    /// <param name="left">The value which <paramref name="right" /> divides.</param>
+    /// <param name="right">The value which divides <paramref name="left" />.</param>
+    /// <returns>The quotient and remainder of <paramref name="left" /> divided-by <paramref name="right" />.</returns>
+    static virtual (TVector Quotient, TVector Remainder) DivRem(TVector left, TVector right);
+
+    /// <summary>Computes the number of bits that are set in a value.</summary>
+    /// <param name="value">The value whose set bits are to be counted.</param>
+    /// <returns>The number of set bits in <paramref name="value" />.</returns>
+    static abstract TVector PopCount(TVector value);
+
+    /// <summary>Tries to read a two's complement number from a span, in big-endian format, and convert it to an instance of the current type.</summary>
+    /// <param name="source">The span from which the two's complement number should be read.</param>
+    /// <param name="isUnsigned"><c>true</c> if <paramref name="source" /> represents an unsigned two's complement number; otherwise, <c>false</c> to indicate it represents a signed two's complement number.</param>
+    /// <param name="value">On return, contains the value read from <paramref name="source" /> or <c>default</c> if a value could not be read.</param>
+    /// <returns><c>true</c> if the value was succesfully read from <paramref name="source" />; otherwise, <c>false</c>.</returns>
+    static abstract bool TryReadBigEndian(ReadOnlySpan<byte> source, bool isUnsigned, [MaybeNullWhen(false)] out TVector value);
```

Comment:
How will this work? For scalars, it reads the entire `source` as one scalar `T` and returns false if its too big.

For a vector, you can't know where one T stops and the next begins.